### PR TITLE
`Forms` : Provide a custom barcode scanner action

### PR DESCRIPTION
**Pull request by kaushikrw**
Related to issue: #[apollo/924](https://devtopia.esri.com/runtime/apollo/issues/924)

### Description:

Adds a function parameter to the `FeatureForm` component that can be used to listen to any taps on barcode field. This is used to enable a custom barcode scanner implementation.

### Summary of changes:

- Feature Addition: Implemented custom barcode click actions within FeatureForm.  
  - A new parameter `onBarcodeButtonClick: ((FieldFormElement) -> Unit)? = null` is provided on the `FeatureForm`  composable.
  - The `FieldFormElement` object can be stored and after a successful scan `FieldFormElement.updateValue` should be called to update the field value. This will automatically cause a recomposition and any changes are visible in the form.
  - When this is set to null (default), the internal MLKit based barcode scanner is used. If the callback is instead utilized, the custom action is invoked when the user taps on the barcode accessory.
  - Added internal functionality to support this custom barcode click actions for `FieldFormElement` and `GroupFormElement`.
  - ~~Note that this will break binary compatibility but source compatibility is preserved.~~.
  - Added overloaded methods with hidden deprecations to preserve binary compatibility. [Ref](https://github.com/androidx/androidx/blob/androidx-main/compose/docs/compose-component-api-guidelines.md#evolution-of-the-component-apis).

- UI Tests: Added UI tests to verify the display and interaction of barcode form elements.  
  - Added test to verify a provided custom action.
  - Added test to verify the default action in case none is provided.

- Documentation: Updated documentation to reflect the new feature.
- Updated API dump.

### Pre-merge Checklist

- a [vTest](https://runtime-kotlin.esri.com/view/all/job/vtest/job/toolkit/) Job for this PR has been run
  - [x] link: https://runtime-kotlin.esri.com/job/vtest/job/toolkit/180/Test_20Report_20-_20Integration/
  - [x] https://runtime-kotlin.esri.com/job/vtest/job/toolkit/185/
- Unit and/or integration tests have been added to exercise this PR's logic, and the tests are passing:
  - [x] Yes
  - [ ] No

### CODE REVIEW

File: toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/FeatureForm.kt [hunk 1]
```diff
@@ -161,7 +169,8 @@ public fun FeatureForm(
     modifier: Modifier = Modifier,
     validationErrorVisibility: ValidationErrorVisibility = ValidationErrorVisibility.Automatic,
     colorScheme: FeatureFormColorScheme = FeatureFormDefaults.colorScheme(),
-    typography: FeatureFormTypography = FeatureFormDefaults.typography()
+    typography: FeatureFormTypography = FeatureFormDefaults.typography(),
+    onBarcodeAccessoryClick: ((FieldFormElement) -> Unit)? = null
```
Comment (kaushikrw) [inline toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/FeatureForm.kt:173]: Is there a better name for this parameter?

File: toolkit/featureforms/api/featureforms.api [hunk 2]
```diff
@@ -1,5 +1,5 @@
 public final class com/arcgismaps/toolkit/featureforms/FeatureFormKt {
-	public static final fun FeatureForm (Lcom/arcgismaps/mapping/featureforms/FeatureForm;Landroidx/compose/ui/Modifier;Lcom/arcgismaps/toolkit/featureforms/ValidationErrorVisibility;Lcom/arcgismaps/toolkit/featureforms/theme/FeatureFormColorScheme;Lcom/arcgismaps/toolkit/featureforms/theme/FeatureFormTypography;Landroidx/compose/runtime/Composer;II)V
+	public static final fun FeatureForm (Lcom/arcgismaps/mapping/featureforms/FeatureForm;Landroidx/compose/ui/Modifier;Lcom/arcgismaps/toolkit/featureforms/ValidationErrorVisibility;Lcom/arcgismaps/toolkit/featureforms/theme/FeatureFormColorScheme;Lcom/arcgismaps/toolkit/featureforms/theme/FeatureFormTypography;Lkotlin/jvm/functions/Function1;Landroidx/compose/runtime/Composer;II)V
```
Comment (sorenoid) [inline toolkit/featureforms/api/featureforms.api:2]: possibly better to deprecate the old function and add an overload. We will discuss tomorrow at coffee meeting.

File: toolkit/featureforms/src/androidTest/java/com/arcgismaps/toolkit/featureforms/BarcodeTests.kt [hunk 3]
```diff
@@ -61,4 +73,130 @@ class BarcodeTests : FeatureFormTestRunner(
...
+        // Check the scan icon is displayed
+        val scanIcon = barcodeFormElement.onChildWithContentDescription("scan barcode")
+        scanIcon.assertIsDisplayed()
+        scanIcon.performClick()
+        // Check the default barcode scanner is displayed
+        val scanner = composeTestRule.onNodeWithContentDescription("MLKit Barcode Scanner")
+        scanner.assertIsDisplayed()
+    }
+
+    @Test
```
Comment (sorenoid) [inline toolkit/featureforms/src/androidTest/java/com/arcgismaps/toolkit/featureforms/BarcodeTests.kt:183]: Missing given/when/then

File: toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/FeatureForm.kt (hunk 1 above)
Comment (sorenoid) [reply toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/FeatureForm.kt:173]: I prefer `...Tap` but not strongly.

**Review by sorenoid — APPROVED** [review]
Pending our discussion about adding parameters to released composable functions, I just have some minor feedback. Looks great

File: toolkit/featureforms/src/androidTest/java/com/arcgismaps/toolkit/featureforms/BarcodeTests.kt (hunk 3 above)
Comment (kaushikrw) [reply toolkit/featureforms/src/androidTest/java/com/arcgismaps/toolkit/featureforms/BarcodeTests.kt:188]: added in https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/pull/639/commits/283eb55aef7f9a606c41a323c8e9def655baa3fd

File: toolkit/featureforms/api/featureforms.api (hunk 2 above)
Comment (kaushikrw) [reply toolkit/featureforms/api/featureforms.api:4]: I've added hidden deprecations for these.

File: toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/FeatureForm.kt (hunk 1 above)
Comment (kaushikrw) [reply toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/FeatureForm.kt]: Renamed to `onBarcodeButtonClick`.

**Comment by kaushikrw** [comment]
Thanks @sorenoid. I've updated the PR with the changes.

File: toolkit/featureforms/api/featureforms.api [hunk 4]
```diff
@@ -207,6 +209,7 @@ public final class com/arcgismaps/toolkit/featureforms/theme/FeatureFormTypograp
 
 public final class com/arcgismaps/toolkit/featureforms/theme/GroupElementColors {
 	public static final field $stable I
+	public synthetic fun <init> (JJJJJLkotlin/jvm/internal/DefaultConstructorMarker;)V
```
Comment (kaushikrw) [inline toolkit/featureforms/api/featureforms.api:212]: Missed generating this in https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/pull/642

**Review by sorenoid — APPROVED** [review]

**Pull request merged**